### PR TITLE
Introduce physics interface

### LINKED ISSUE #8: Second stage refactor: improve physics interface
Currently, the physics interface is fairly basics and provides everything in one single `PoissonPhysics` class which provides the integrand of the potential energy of a Poisson's problem. This is done in each script or integration test. The goal of the issue is to provide an interface to generalize this approach. Ideally, we want to be able to combine the different physics blocks and add/subtract them at will to form physical equations. For Poisson equation this could look like in `main()`:
```python
    u = Field("displacement", mesh, dirichlet_nodes=[0, N - 1])
    physics = StrainEnergyDensity(u, u) # Provide 0.5 * du_dx**2. Can be used for 0.5 * du_dx * dv_dx in case the two parameters are u,v
              - LinearSourceTerm(f, u) # Provide f * u
```

This interface should also introduce more complex integration tests with less trivial physics terms.

### PR DESCRIPTION
**Pull request by jdichamp**
**IMPORTANT**: should be reviewed and merged after #11. 

This PR provides an interface to combine `Term`s to build a physical based loss. To do so, a `FieldLayout` is introduced and a late retrieval of the actual interpolated fields is done when integrating all the terms in the loss. In the process a `jacobian_field()` function is provided that works in all dimensions as well as a `inner()` function for inner product. Those will be later be moved to a `math` package. Added more tests (#5) in general and all new functionality should be tested.  

Closes #8

### CODE REVIEW

File: scripts/design_poc/main.py [hunk 1]
```diff
@@ -72,24 +96,25 @@ def main():
...
+        loss.backward(retain_graph=True)
         optimizer.step()
 
         loss_history.append(loss.item())
         print(f"{i=} loss={loss.item():.3e}", end="\r")
 
     print("\n* Evaluation")
     # At quadrature points
-    x_q, u_q, _ = model.evaluator.evaluate()
+    result = field_layout["displacement"]
```
Comment (AlexandreDabySeesaram) [inline scripts/design_poc/main.py:126]: I really like the abstraction level here

File: src/neurom/physics_loss.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    This integrates the terms present in the ``physics`` Term with the fields in the provided ``field_layout``.
+
+    Args:
+        physics (Term): The physics providing the integrand.
+        field_layout (FieldLayout): The field layout in which we will look for the actual fields in the integrand.
+    Attributes:
+        physics (Term): The physics providing the integrand.
+        field_layout (FieldLayout): The field layout in which we will look for the actual fields in the integrand.
+    """
+
```
Comment (AlexandreDabySeesaram) [inline src/neurom/physics_loss.py:21]: We might want to 

1. Split the integrations of the integrand based on the required quadrature
2. Or use several PhysicsLoss for different quadratures

**Review by AlexandreDabySeesaram — APPROVED** [review]
